zetasql/fuzzing/zetasql_expression_proto_to_string.cc

```diff
@@ -0,0 +1,132 @@
+//
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <string>
+
+#include "zetasql/fuzzing/common.h"
+#include "zetasql/fuzzing/zetasql_expression_grammar.pb.h"
+
+using namespace zetasql_expression_grammar;
+
+namespace zetasql_fuzzer {
+
+std::string HandleDefault() {
+  return "0";
+}
+
+// Forward declaration
+CONV_FN(Expression, expr);
+
+CONV_FN(LiteralExpr, lit_expr);
+CONV_FN(IntegerLiteral, integer);
+CONV_FN(NumericLiteral, numeric);
+
+CONV_FN(CompoundExpr, comp_expr);
+CONV_FN(AdditiveOperation, additive);
+CONV_FN(MultiplicativeOperation, multiplicative);
+
+CONV_FN(Expression, expr) {
+  using ExprType = Expression::ExprOneofCase;
+  switch (expr.expr_oneof_case()) {
+    case ExprType::kLiteral:
+      return LiteralExprToString(expr.literal());
+    case ExprType::kExpr:
+      return CompoundExprToString(expr.expr());
+    default:
+      return HandleDefault();
+  }
+}
+
+CONV_FN(LiteralExpr, literal) {
+  using LitExprType = LiteralExpr::LiteralOneofCase;
+  std::string ret;
+  switch (literal.literal_oneof_case()) {
+    case LitExprType::kSpecialLiteral:
+      switch (literal.special_literal()) {
+        case LiteralExpr::V_NULL:
+          return "NULL";
+        default:
+          std::abort();
```

Comment:
Consider outputting a logging statement in addition to just crashing here.

Reply:
Would you suggest using `assert(false && "message")`;

Reply:
So there is probably a CHECK macro available through abseil or somewhere else in zetasql.  I'd look for that first.

Reply:
Be careful about asserts, since they sometimes get removed in production.

I usually use LOG(FATAL) << "message"; but I don't know if you have access to that logging here.

Reply:
@shumway so likely we don't have logging function in OSS-Fuzz as it's not provided in the interface. Currently fuuzers are *always* compiled compiled with Bazel debug mode on so I'm not too worried about it. But I agree that maybe some kind of macro may help.